### PR TITLE
Add a setter method variable_scope.auto_reuse_variables() to enable AUTO_REUSE

### PR DESCRIPTION
**Pull request by bowang**
Currently we can enable variable reuse in a scope with variable_scope.reuse_variables(). However, there is no handy method to enable auto reuse yet. This PR adds this functionality.

### CODE REVIEW

**Comment by asimshankar** [comment]
(From API review)

We'd like to avoid this change as it makes the variable scope harder to reason about (harder than it is now :). We don't have a way to set `self._reuse = False` either.

If you have a strong use case for this, we'd love to hear it. Otherwise, we're tempted to drop this change.

Thanks.

**Comment by bowang** [comment]
Thanks for the feedback @asimshankar. I understand the reason for not having an API to reset the reuse status.

Currently the only way to enable auto-reuse for a scope is at its creation (if I understand it correctly). This forces me to create an auxiliary scope purely for this purpose in multi-GPU training. Models created in the auxiliary scope inherits the scope name as prefix, which causes the parameter names to be different from those in the pre-trained models without this scope. Even though I can manually match the names, it would be much easier and elegant to go without the auxiliary scope.

**Comment by asimshankar** [comment]
@bowang : Could you expand a bit on the code you're writing? `AUTO_REUSE` is a relatively new thing anyway. The idea is that you'd create variables once (so with `reuse=False`) and then create more operations (e.g., towers in a multi-GPU configuration) with `reuse=True`. Can you use that pattern in your code?

**Comment by bowang** [comment]
@asimshankar Here are some example code with/without `AUTO_REUSE`

Without `AUTO_REUSE`
```
for i in range(num_gpus):
  with tf.device('/gpu:%d' % i):
    with tf.variable_scope('gpu', reuse=(i>0)):
      create_gpu_tower(i)
```

With `AUTO_REUSE`
```
tf.get_variable_scope()._reuse = tf.AUTO_REUSE
for i in range(num_gpus):
  with tf.device('/gpu:%d' % i):
    create_gpu_tower(i)
```

By using `AUTO_REUSE`, we can avoid create the auxiliary scope `gpu`, which makes it easier and compatible with models created for a single GPU.

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @caisq: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @caisq: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @caisq: It has been 37 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @caisq: It has been 52 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @caisq: It has been 67 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @caisq: It has been 82 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @caisq: It has been 97 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @caisq: It has been 112 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @caisq: It has been 127 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @caisq: It has been 142 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @caisq: It has been 157 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @caisq: It has been 172 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @caisq: It has been 187 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @caisq: It has been 202 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by asimshankar** [comment]
As we make progress on TF 2.0, variable_scope becomes obsolete (and instead sharing is achieved by sharing the Python `Variable` object). Hence closing this PR.
